src/linker/Linker/LinkContext.cs

```diff
@@ -583,6 +602,22 @@ public void LogWarning (string text, int code, IMemberDefinition origin, int? il
 			LogWarning (text, code, _origin, subcategory);
 		}
 
+		/// <summary>
+		/// Display a warning message to the end user.
+		/// This API is used for warnings defined in the linker, not by custom steps. Warning
+		/// versions are inferred from the code, and every warning that we define is versioned.
+		/// </summary>
+		/// <param name="text">Humanly readable message describing the warning</param>
+		/// <param name="code">Unique warning ID. Please see https://github.com/dotnet/linker/blob/main/docs/error-codes.md for the list of warnings and possibly add a new one</param>
+		/// <param name="origin">Type or member where the warning is coming from</param>
+		/// <param name="subcategory">Optionally, further categorize this warning</param>
+		/// <returns>New MessageContainer of 'Warning' category</returns>
+		public void LogWarning (string text, DiagnosticId code, IMemberDefinition origin, int? ilOffset = null, string subcategory = MessageSubCategory.None)
```

Comment:
Why do you need a string `text` if all the time you do
`new DiagnosticString (code).GetMessage (some other values)` ?

Reply:
The return value of `new DiagnosticString (code).GetMessage (args)` is a string. GetMessage method takes the DiagnosticString name and looks in SharedStrings.resx file for an entry with `name+Message` then substitutes the args into the SharedStrings.resx value and returns the formed string.

Reply:
I think the idea is to have an overload which takes `DiagnosticId` and "params" and internally it does the `new DiagnosticString(code).GetMessage(params)` for you. The point is that almost all call sites in the linker now have the `DiagnosticString(code).GetMessage(params)` code, so to remove that duplication, make it part of the `LogWarning`. Ideally the call site should only pass in things in their simplest form and as few as viable. Currently that's not the case.